### ISSUE TITLE
Release generation for specific pg version

### ISSUE THREAD

**Issue by ccleve** [zombodb/pgx#288]
```
cargo pgx package pg13
error: Found argument 'pg13' which wasn't expected, or isn't valid in this context

cargo pgx package --features pg13
building extension with features `pg13 pg14`
"cargo" "build" "--release" "--features" "pg13 pg14" "--no-default-features"

```
generates a lengthy error trace, which, oddly, I can't seem to capture to a file. It's mostly about ambiguous symbols, which makes sense if it's trying to do pg13 and pg14 at the same time.

`"cargo" "build" "--release" "--features" "pg13" "--no-default-features"`

Does not crash, but does not create a release, either.

Document how to create a release for a particular pg version.

**Comment by eeeebbbbrrrr** [comment]
https://github.com/zombodb/pgx/blob/master/cargo-pgx/README.md#building-an-installation-package

Note that `cargo pgx package` uses the `pg_config` on your $PATH, **not** any of the versioned postgres instances pgx manages for you during development.

This is so that files are written to the proper directory structure for easy tarball/rpm/deb/apk packaging, which you'd need to do yourself.

**Comment by ccleve** [comment]
I get it. It uses pg_config to figure out your local pg install, and then build for that specific pg version and installation path(s).

If you want to use CI to build for multiple pg versions and OSes, would it not make sense to build against the pgx-managed versions, and put your binary/.sql/.control files in fixed locations in, say, a .zip or .tar.gz file?

This way it might be easier to create rpm/deb/etc packages. I don't know, though, not having done it before. Just a thought for your consideration.

**Comment by eeeebbbbrrrr** [comment]
> If you want to use CI to build for multiple pg versions and OSes, would it not make sense to build against the pgx-managed versions, and put your binary/.sql/.control files in fixed locations in, say, a .zip or .tar.gz file?

No.  Because the pgx-managed versions of Postgres, if downloaded/compiled by pgx, default to `~/.pgx/`, which is entirely different from what Linux Distros do for Postgres (and they all do something different, of course).

That said, if you look at what @zombodb does, it's quite complex and it does lean heavily on distro-managed Postgres versions.  https://github.com/zombodb/zombodb/tree/master/docker-build-system

Postgres extension packaging is really quite a pain because ya just don't know what all the various distros do or might decide to do tomorrow.

**Comment by ccleve** [comment]
Man, this really is a nightmare. There's got to be a better way.

Maybe we need an extension to manage extensions.

I'll take another crack at this in a few months when my project is ready to go.

**Comment by eeeebbbbrrrr** [comment]
I think “nightmare” is a bit harsh, but I feel your pain. That’s why I invested so much energy into ZDB’s build system stuff.  And why I bought a $5k threadripper to run it.   And then charge $75/mo for binary artifact downloads. It’s actual work!

There’s really two problems here and I bet you haven’t discovered the other one yet — extension schema version upgrades/migrations.  We have lots of thoughts on how to make pgx automate this, and it definitely needs to since it makes the schema, but it’s gonna be a lot of work. 

The other problem is what we’re discussing here, which is pure packaging. I don’t have any brilliant ideas at the moment, but it’s something I stare off into space and think about often. 

I’m gonna re-open this cuz others I’m sure will have some thoughts too.

**Comment by ccleve** [comment]
Ok, so long as you're inviting thoughts:

PGXN has a zip file format for extensions. Seems to work ok. They just bypass package managers. You have to install the pgxn client, but then you just say `pgxn install semver` and you're good.

The downside is that the pgxn install process compiles source and chokes if the toolchain is incomplete. This isn't good, but I think the idea of a simple command-line client to install extensions is the way to go.

Installing from source is also a bad idea because Rust compile times are way too high, particularly if you also have to install cargo-pgx, which in turn requires the C toolchain and maybe other stuff. (I just got bit by that. On Ubuntu you have to `apt install pkg-config` in addition to the tools required to compile pg.)

Binaries are better but present other problems. If you install binaries, then you'll have to compile for every unique combination of [extension version] / [pg version] / [OS] / [architecture (x86_64, AMD, etc.)]. This could be a large number of binaries, although I found some suggestions online that Rust binaries could be portable across *some* versions of Linux. So, you'd have to figure out the minimum number of binaries that could handle everything.

Then, if you come up with a common extension .zip file format, things are easy. Just have your custom extension-installer client detect the current OS and architecture, and then have it invoke pg_config to get the destination folders. Copy the extension files and you're done.

So, how to create all these binaries? The options are 1) create some docker-based monstrosity, or 2) figure out how to get TravisCI or Github actions to do the builds on their own hardware, or 3) see if rustup has some cross-compiler abilities (I read that it did), or 4) create a service that does this on your own hardware.

Option 4 is interesting. It could become the successor to pgxn. It could accept commercial extensions and charge for stuff.

Another idea: Option 4 could work in a lazy way. You could say `my_extension_installer install my_extension_name` and it could pull the source on the fly from Github, compile it just for that one OS/architecture, and install it. It might be fast enough on server hardware. It would then cache that binary in S3 or someplace.

This would be really cool if it became the standard way to install all extensions, including PostGIS. In a happy, happy world the client would get bundled with Postgres itself (though I wouldn't hold my breath).

**Comment by ccleve** [comment]
@eeeebbbbrrrr I just stumbled across this: https://pgt.dev/ Wonder if it could be a universal solution to the deployment problem?

**Comment by eeeebbbbrrrr** [comment]
> @eeeebbbbrrrr I just stumbled across this: https://pgt.dev/ Wonder if it could be a universal solution to the deployment problem?

Probably.  I'm friendly with those folks and I expect they'll have support for pgrx-based extensions after awhile.  A handful of the company's own extensions are written with pgrx, so they'll need it anyways.
